File: 01-git/mkdir-erdman/mkdir.go
package main

import (
	"crypto/sha256"
	"fmt"
	"os"
	"sort"
	"strings"
)

func ParseArgs() (string, string) {
	args := os.Args[1:]
	if len(args) != 2 {
		fmt.Println("Usage: ./mkdir <path> <hash>")
		os.Exit(1)
	}
	return args[0], args[1]
}

func MakeDir(path string, hash string) string {
	list := GetFileList(hash)
	var filteredList []string
	for _, elem := range list {
		if !strings.Contains(elem, ".parent/") && elem != "" {
			filteredList = append(filteredList, elem)
		}
	}
	list = filteredList

	newDirHash := WriteBlob("")
	list = append(list, fmt.Sprintf(".parent/\t%s", hash))

	fullPath := []string{}
	dirList := strings.Split(path, "/")

	if strings.Contains(path, "/") {
		parentDir := hash
		for i := range dirList {
			fullPath = append(fullPath, parentDir)
			if i+1 == len(dirList) {
				f, err := os.OpenFile(parentDir, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
				CheckError(err)
				defer f.Close()
				//f.WriteString(fmt.Sprintf("%s/\t%s\n", dirList[i], newDirHash))
			} else {
				data, err := os.ReadFile(parentDir)
				CheckError(err)
				if strings.Contains(string(data), dirList[i]) {
					parentList := GetFileList(parentDir)
					found := false
					for _, f := range parentList {
						if strings.Contains(f, dirList[i]) {
							parentDir = strings.Split(f, "\t")[1]
							found = true
							break
						}
					}
					if found {
						continue
					}
					fmt.Printf("incorrect structure\n")
					os.Exit(1)
				} else {
					fmt.Printf("%s doesn't exist\n", dirList[i])
					os.Exit(1)
				}
			}
		}
	} else {
		list = append(list, fmt.Sprintf("%s/\t%s", path, newDirHash))
	}

	if len(fullPath) > 1 {
		updHash := newDirHash
		updDirIndex := len(dirList) - 1
		updDir := dirList[updDirIndex]
		for i := len(fullPath) - 1; i >= 0; i-- {
			prevList := GetFileList(fullPath[i])
			updated := false
			for j := range prevList {
				if strings.Contains(prevList[j], updDir) {
					updated = true
					prevList[j] = fmt.Sprintf("%s/\t%s", updDir, updHash)
				}
				if i == 0 && strings.Contains(prevList[j], ".parent/") {
					prevList[j] = fmt.Sprintf(".parent/\t%s", hash)
				}
			}
			if !updated {
				prevList = append(prevList, fmt.Sprintf("%s/\t%s", updDir, updHash))
			}

			sort.Strings(prevList)
			resStr := ""
			for _, el := range prevList {
				resStr += el
				resStr += "\n"
			}
			updHash = WriteBlob(resStr)

			updDirIndex--
			if updDirIndex >= 0 {
				updDir = dirList[updDirIndex]
			}
		}
		return updHash
	}

	sort.Strings(list)
	newDirData := strings.Join(list, "\n") + "\n"
	newRootHash := Sha256(newDirData)
	err := os.WriteFile(newRootHash, []byte(newDirData), 0666)
	CheckError(err)
	return newRootHash
}

func GetFileList(path string) []string {
	data, err := os.ReadFile(path)
	CheckError(err)
	list := strings.Split(string(data), "\n")
	resList := []string{}
	for _, el := range list {
		if el != "" {
			resList = append(resList, el)
		}
	}
	return resList
}

func Sha256(s string) string {
	h := sha256.New()
	h.Write([]byte(s))
	return fmt.Sprintf("%x", h.Sum(nil))
}

func WriteBlob(data string) string {
	hash := Sha256(data)
	err := os.WriteFile(hash, []byte(data), 0666)
	CheckError(err)
	return hash
}

func CheckError(err error) {
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
}

func main() {
	path, hash := ParseArgs()
	resultHash := MakeDir(path, hash)
	fmt.Printf("New hash: %s\n", resultHash)
}

File: 01-git/mkdir-mukhin/mkdir.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <openssl/sha.h>
#include <cstring>

std::string Sha256(const std::string &data) {
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char *>(data.c_str()), data.size(), hash);
    std::stringstream hexStream;
    for (int i = 0; i < SHA256_DIGEST_LENGTH; ++i) {
        hexStream << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(hash[i]);
    }
    return hexStream.str();
}

std::string makeBlob(const std::string &data) {
    std::string hash = Sha256(data);
    std::ofstream out(hash, std::ofstream::binary);
    if (!out.is_open()) {
        std::cerr << "Failed to open file for writing" << std::endl;
        exit(1);
    }
    out << data;
    return hash;
}

std::vector<std::string> getList(const std::string &path) {
    std::ifstream file(path);
    if (!file.is_open()) {
        std::cerr << "Failed to open file for reading" << std::endl;
        exit(1);
    }

    std::vector<std::string> resList;
    std::string line;
    while (std::getline(file, line)) {
        if (!line.empty()) {
            resList.push_back(line);
        }
    }
    return resList;
}

std::pair<std::string, std::string> ParseArgs(int argc, char *argv[]) {
    if (argc != 3) {
        std::cerr << "Wrong input: ./mkdir <path> <hash>";
        exit(1);
    }
    return {argv[1], argv[2]};
}

std::vector<std::string> SplitPath(const std::string &path) {
    std::vector<std::string> list;
    size_t start = 0;
    size_t end = path.find("/");
    while (end != std::string::npos) {
        list.push_back(path.substr(start, end - start));
        start = end + 1;
        end = path.find("/", start);
    }
    list.push_back(path.substr(start, end));
    return list;
}

std::string CreateData(std::vector<std::string> &list) {
    std::sort(list.begin(), list.end());
    std::string data;
    for (const auto &item: list) {
        data += item + "\n";
    }
    return data;
}

std::string CreateNewHash(std::vector<std::string> &list) {
    std::string newData = CreateData(list);
    std::string newHash = Sha256(newData);
    std::ofstream(newHash) << newData;
    return newHash;
}

std::string UpdateHash(std::vector<std::string> &path, std::vector<std::string> &dirList,
                       std::string hashToUpd, const std::string &hash) {
    int updIndex = dirList.size() - 1;
    std::string updDir = dirList[updIndex];
    for (int i = path.size() - 1; i >= 0; --i) {
        auto lastList = getList(path[i]);
        bool updated = false;
        for (size_t j = 0; j < lastList.size(); ++j) {
            if (lastList[j].find(updDir) != std::string::npos) {
                lastList[j] = updDir + "/\t" + hashToUpd;
                updated = true;
            }
            if (lastList[j].find(".parent/") != std::string::npos && i == 0) {
                lastList[j] = ".parent/\t" + hash;
            }
        }
        if (!updated) {
            lastList.push_back(updDir + "/\t" + hashToUpd);
        }

        std::string data = CreateData(lastList);
        hashToUpd = makeBlob(data);

        --updIndex;
        if (updIndex >= 0) {
            updDir = dirList[updIndex];
        }
    }
    return hashToUpd;
}

std::string MakeDir(const std::string &path, const std::string &hash) {
    auto list = getList(hash);
    std::vector<std::string> clearList;
    for (const auto &elem: list) {
        if (elem.find(".parent/") == std::string::npos && !elem.empty()) {
            clearList.push_back(elem);
        }
    }
    list = std::move(clearList);

    std::string newHash = makeBlob("");
    list.push_back(".parent/\t" + hash);

    std::vector<std::string> fullPath;
    auto dirList = SplitPath(path);

    if (path.find("/") == std::string::npos) {
        list.push_back(path + "/\t" + newHash);
    } else {
        std::string parentDir = hash;
        for (size_t i = 0; i < dirList.size(); ++i) {
            fullPath.push_back(parentDir);
            if (i + 1 == dirList.size()) {
                std::fstream f(parentDir, std::fstream::app | std::fstream::out);
                if (!f.is_open()) {
                    std::cerr << "File open failed" << std::endl;
                    exit(1);
                }
            } else {
                std::ifstream file(parentDir);
                if (!file.is_open()) {
                    std::cerr << "File open failed" << std::endl;
                    exit(1);
                }
                std::string data((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
                if (data.find(dirList[i]) != std::string::npos) {
                    auto parentList = getList(parentDir);
                    bool found = false;
                    for (const auto &f: parentList) {
                        if (f.find(dirList[i]) != std::string::npos) {
                            parentDir = f.substr(f.find("\t") + 1);
                            found = true;
                            break;
                        }
                    }
                    if (found) {
                        continue;
                    }

                    std::cerr << "Incorrect structure\n";
                    exit(1);
                } else {
                    std::cerr << dirList[i] << " doesn't exist\n";
                    exit(1);
                }
            }
        }
    }

    if (fullPath.size() <= 1) {
        return CreateNewHash(list);
    }
    return UpdateHash(fullPath, dirList, newHash, hash);
}

int main(int argc, char *argv[]) {
    auto [path, hash] = ParseArgs(argc, argv);
    std::string resultHash = MakeDir(path, hash);
    std::cout << "New hash: " << resultHash << std::endl;
    return 0;
}